pkg/lib/cockpit-components-terminal.tsx

```diff
@@ -120,71 +226,92 @@
         this.getText = this.getText.bind(this);
         this.setTerminalTheme = this.setTerminalTheme.bind(this);
 
-        const term = new Term({
-            cols: props.cols || 80,
-            rows: props.rows || 25,
-            screenKeys: true,
-            cursorBlink: true,
-            fontSize: props.fontSize || 16,
-            fontFamily: 'Menlo, Monaco, Consolas, monospace',
-            screenReaderMode: true,
-            showPastingModal: false,
-        });
+        if (this.props.state) {
+            cockpit.assert(!this.props.channel);
+            this.terminal_state = this.props.state;
+        } else {
+            cockpit.assert(this.props.channel);
+            this.terminal_state = new TerminalState(this.props.channel);
+        }
 
-        this.terminalRef = React.createRef();
+        const term = this.terminal_state.terminal;
 
-        term.onData(function(data) {
-            if (this.props.channel.valid) {
-                /* HACK: Ctrl+Space (and possibly other characters) is a disaster: While it is U+00A0 in unicode, with
-                 * an UTF-8 representation of 0xC2A0, the "visible" string in JS is 0x00 (with TextEncoder,
-                 * btoa(), and string comparison). The internal representation retains half of it, and trying to send
-                 * it to the websocket would result in a single 0xA0, which is invalid UTF-8 (and causes the session
-                 * to crash). So intercept and ignore such broken chars.
-                 * See https://github.com/cockpit-project/cockpit/issues/21213 */
-                if (data === '\x00') {
-                    console.log("terminal: ignoring invalid input", data);
-                    return;
-                }
-                this.props.channel.send(data);
-            }
-        }.bind(this));
+        this.terminalRef = React.createRef<HTMLDivElement>();
 
         if (props.onTitleChanged)
             term.onTitleChange(props.onTitleChanged);
 
         this.terminal = term;
         this.state = {
             showPastingModal: false,
-            cols: props.cols || 80,
-            rows: props.rows || 25
+            cols: term.cols,
+            rows: term.rows
         };
     }
 
-    componentDidMount() {
-        this.terminal.open(this.terminalRef.current);
-        this.terminal.loadAddon(new CanvasAddon());
-        this.connectChannel();
+    mountTerminal(state: TerminalState) {
+        this.terminal = state.terminal;
+        this.terminalRef.current?.appendChild(state.element);
+        this.terminal.open(state.element);
+
+        if (this.props.fontSize)
+            this.terminal.options.fontSize = this.props.fontSize;
+
+        if (this.props.cols && this.props.rows) {
+            this.resizeTerminal(this.props.cols, this.props.rows);
+        }
+
+        this.setTerminalTheme(this.props.theme || 'black-theme');
```

Comment:
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-21680-4a6f1830-20250815-070855-fedora-42-devel/Coverage/lcov/github-pr.diff.gcov.html)